akka-sample-kafka-to-sharding/processor/src/main/scala/sample/sharding/kafka/UserEventsKafkaProcessor.scala

```diff
@@ -0,0 +1,92 @@
+package sample.sharding.kafka
+
+import akka.Done
+import akka.pattern.retry
+import akka.actor.ActorSystem
+import akka.actor.Scheduler
+import akka.actor.typed.Behavior
+import akka.actor.typed.scaladsl.AskPattern._
+import akka.actor.typed.scaladsl.Behaviors
+import akka.actor.typed.scaladsl.adapter._
+import akka.kafka.ConsumerSettings
+import akka.kafka.Subscriptions
+import akka.kafka.scaladsl.Consumer
+import akka.stream.scaladsl.Sink
+import akka.stream.scaladsl.Source
+import akka.util.Timeout
+import org.apache.kafka.clients.consumer.ConsumerConfig
+import org.apache.kafka.clients.consumer.ConsumerRecord
+import org.apache.kafka.common.serialization.ByteArrayDeserializer
+import org.apache.kafka.common.serialization.StringDeserializer
+import sample.sharding.kafka.serialization.UserPurchaseProto
+
+import scala.concurrent.ExecutionContextExecutor
+import scala.concurrent.Future
+import scala.concurrent.duration._
+import scala.util.Try
+
+object UserEventsKafkaProcessor {
+
+  sealed trait Command
+  private case class KafkaConsumerStopped(reason: Try[Any]) extends Command
+
+  def apply(): Behavior[Nothing] = {
+    Behaviors
+      .setup[Command] { ctx =>
+        val processorSettings = ProcessorConfig(ctx.system.settings.config.getConfig("kafka-to-sharding-processor"))
+        implicit val classic: ActorSystem = ctx.system.toClassic
+        implicit val ec: ExecutionContextExecutor = ctx.executionContext
+        implicit val scheduler: Scheduler = classic.scheduler
+        // TODO config
+        val timeout = Timeout(3.seconds)
+        val rebalancerRef = ctx.spawn(TopicListener(UserEvents.TypeKey), "rebalancerRef")
+        val shardRegion = UserEvents.init(ctx.system)
+        val consumerSettings =
+          ConsumerSettings(ctx.system.toClassic, new StringDeserializer, new ByteArrayDeserializer)
+            .withBootstrapServers(processorSettings.bootstrapServers)
+            .withGroupId(processorSettings.groupId)
+            .withProperty(ConsumerConfig.AUTO_OFFSET_RESET_CONFIG, "earliest")
+            .withStopTimeout(0.seconds)
+
+        val subscription = Subscriptions.topics(processorSettings.topic).withRebalanceListener(rebalancerRef.toClassic)
+
+        val kafkaConsumer: Source[ConsumerRecord[String, Array[Byte]], Consumer.Control] =
+          Consumer.plainSource(consumerSettings, subscription)
```

Comment:
Not a big deal for a sample, but if we want at-least-once guarantees we should use a `committableSource` and then a `Committer.sink|flow` after the processing stage.

Maybe this could be a TODO if/when this sample is updated for reliable delivery, or sharding at-least-once.

Reply:
Added a TODO